### DISCUSSION TITLE
I am having problems because FastAPI is doing too much unskippable "magic" behind the scenes while returning response.

### DISCUSSION THREAD

**Discussion by caner-cetin** [fastapi/fastapi#10027] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code
Here is a example with custom ObjectID type that I just copied from somewhere. I cant share any more than this MRE, because, you know, NDA.
```python


from typing import Any, Callable
from typing_extensions import Annotated

from fastapi import FastAPI, Response
from bson import ObjectId
from pydantic import (
    BaseModel,
    Field,
    GetJsonSchemaHandler,
)

from pydantic.json_schema import JsonSchemaValue
from pydantic_core import core_schema


class _ObjectIdPydanticAnnotation:
    @classmethod
    def __get_pydantic_core_schema__(
        cls,
        _source_type: Any,
        _handler: Callable[[Any], core_schema.CoreSchema],
    ) -> core_schema.CoreSchema:
        def validate_from_str(id_: str) -> ObjectId:
            return ObjectId(id_)

        from_str_schema = core_schema.chain_schema(
            [
                core_schema.str_schema(),
                core_schema.no_info_plain_validator_function(validate_from_str),
            ]
        )

        return core_schema.json_or_python_schema(
            json_schema=from_str_schema,
            python_schema=core_schema.union_schema(
                [
                    # check if it's an instance first before doing any further work
                    core_schema.is_instance_schema(ObjectId),
                    from_str_schema,
                ]
            ),
            serialization=core_schema.plain_serializer_function_ser_schema(
                lambda instance: str(instance)
            ),
        )

    @classmethod
    def __get_pydantic_json_schema__(
        cls, _core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
    ) -> JsonSchemaValue:
        # Use the same schema that would be used for `str`
        return handler(core_schema.str_schema())


def new_oid() -> ObjectId:
    return ObjectId()


OID = Annotated[ObjectId, _ObjectIdPydanticAnnotation]


class Model(BaseModel):
    id: OID = Field(default_factory=new_oid, alias="_id")


app = FastAPI()


def return_response(return_data: Any, media_type: str) -> Response:
    return Response(content=return_data, media_type=media_type)


@app.get("/model", response_model=Model)
async def get_model() -> Model:
    return return_response(Model().model_dump(), "application/json")
```

### Description

Navigate to http://127.0.0.1:{whatever_port}/docs#/default/get_model_model_get and just execute.
You will see that 

```py
  File "/home/caner/goofy-ahh/.venv/lib/python3.11/site-packages/starlette/responses.py", line 55, in __init__
    self.body = self.render(content)
                ^^^^^^^^^^^^^^^^^^^^
  File "/home/caner/goofy-ahh/.venv/lib/python3.11/site-packages/starlette/responses.py", line 63, in render
    return content.encode(self.charset)
           ^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'encode'
``` 
Starlette tries to encode the response.

You can just say "Caner, are you f*!@#ing idiot? just return Model()", and yeah, you are right, but I have many more custom types besides this, and in my original code I absolutely have to use `model_dump()` of Pydantic with custom encoders declared in a custom BaseModel.

Why? Because if i return the model directly, Pydantic will try to encode it. and it will use jsonable_encoder, not the custom encoders of the response model. heheh. another magic I dont have control of. :)

So lets take a look at how shitfest this is:

- If I return the output of `model_dump()`, Pydantic tries to encode it.
- If I return the output of `model_dump()` wrapped inside the `Response` Starlette will try to encode it.
- If I return the Model directly, Pydantic will try to encode it with jsonable_encoders that will lead to some utterly stupid errors like `pydantic_core._pydantic_core.PydanticSerializationError: Unable to serialize unknown type: ObjectId('64d0afd54d551a96b5323a8c')`
- If I return the output of `model_dump()` wrapped inside str `str(model_dump())` and give it to the Response, Starlette again will try to encode it, and it cannot parse JSON.

Like I already have the JSON dumped. Why are you still feeding it to encoders? Why you dont allow me to disable this behaviour?

...

Please, if you are making a magic behind the scenes, let us disable it. Do you know why GoLang is so f$#@king popular compared and preferred in backend? No magic behind the scenes, do you want to do something? here is the thing you want, not the thing you want with a random encoder you see.

_rant over, please help_

### Operating System

Linux

### Operating System Details

does it even matter?

![image](https://github.com/tiangolo/fastapi/assets/92731060/c2f07786-d728-4c7d-84e7-ecf345d712e6)

### FastAPI Version

0.101.0

### Pydantic Version

2.1.1

### Python Version

Python 3.11.4

### Additional Context

<3

**Comment by MrNaif2018** [comment]
Actually similar to my issue #9983 as well
Using Annotated to customize serialization works, but sometimes serialization is somewhere deep in pydantic/fastapi/starlette that is outside of control, impossible to customize.

**Comment by caner-cetin** [comment]
@MrNaif2018 s issue is literally same as mine. Guess I will not get a reply for days and days too lol

**Comment by caner-cetin** [comment]
I solved the problem by dumping the model => then dumping the model dictionary as json => then encoding the json as utf-8 bytes.

in starlette's forced encode, you can see the

```py
    def render(self, content: typing.Any) -> bytes:
        if content is None:
            return b""
        if isinstance(content, bytes):
            return content
        return content.encode(self.charset)
``` 
content gets sent without getting encoded if content is in bytes already, but, come on. I just added two extra steps for sending response. json.dumps and an encode of the json. I am not closing the issue, as this is literally a bandaid fix.

**Comment by n8sty** [comment]
Have you tried directly returning a `Response` object (https://fastapi.tiangolo.com/advanced/response-directly/)?
```
@app.get("/model")
async def get_model() -> Model:
    return JsonResponse(content=Model().model_dump())
```

**Comment by caner-cetin** [comment]
Check my minimal reproducable example again, I am returning lowest level Response that FastAPI allows me.

**Comment by iloveitaly** [comment]
@caner-cetin were you able to solve this? I have a very similar issue—trying to get fastapi to properly encode a https://github.com/akhundMurad/typeid-python/

**Comment by YuriiMotov** [comment, marked as answer]
Just pass `bytes` to `Response.content`:

```python
def return_response(return_data: BaseModel | dict | str | bytes, media_type: str) -> Response:
    data: bytes
    if isinstance(return_data, bytes):
        data = return_data
    elif isinstance(return_data, BaseModel):
        data = return_data.model_dump_json().encode()
    elif isinstance(return_data, dict):
        data = json.dumps(return_data).encode()
    elif isinstance(return_data, str):
        data = return_data.encode()
    else:
        raise ValueError("Unsupported type")

    return Response(content=data, media_type=media_type)

@app.get("/model", response_model=Model)
async def get_model() -> Model:
    return return_response(Model(), "application/json")
```
